openverse_catalog/dags/providers/provider_csv_load_scripts/inaturalist/taxa.sql

```diff
@@ -3,51 +3,165 @@
++ add in inaturalist taxa base data, with the category of "Life" excluded from
+  ancestry strings
+*/
+drop table if exists inaturalist.taxa_with_vernacular;
+create table inaturalist.taxa_with_vernacular as
+(
+    with
+    catalog_of_life_names as
+    (
+        select cast(md5(n.scientificname) as uuid) as md5_scientificname,
+            v.name_language,
+            v.taxon_name,
+            row_number() OVER (partition by n.scientificname
+                order by case when v.name_language='eng' then 0 else 1 end,
+                v.taxon_name)
+                as name_rank,
+            count(case when v.name_language='eng' then 1 end)
+                OVER (partition by n.scientificname
+                rows between unbounded preceding and unbounded following)
+                as english_names
+        FROM inaturalist.col_name_usage n
+            INNER JOIN inaturalist.col_vernacular v on v.taxonid = n.id
+        where length(n.id) <= 10
+    ),
+    catalog_of_life as
+    (
+        SELECT
+            md5_scientificname,
+            string_agg(DISTINCT
+                case when name_language = 'eng' then taxon_name end,
+                ', ') name_english,
+            jsonb_agg(DISTINCT cast((taxon_name, 'inaturalist') as openverse_tag))
+                FILTER (where name_rank > english_names) as tags_nonenglish_vernacular
+        FROM catalog_of_life_names
+        where name_rank <= 20
+        group by 1
+    )
+    select
+        taxa.taxon_id,
+        (case when ancestry='48460' then '' else replace(taxa.ancestry,'48460/','') end)
+            as ancestry, --exclude 'Life' from ancestry
+        left(coalesce(
+            catalog_of_life.name_english,
+            manual_name_additions.vernacular_name, --name_manual,
+            taxa.name -- name_inaturalist,
+        ), 5000) as title,
+        (case when catalog_of_life.name_english is not null or
+            manual_name_additions.vernacular_name is not null
+            then to_jsonb(array_fill(cast((taxa.name, 'inaturalist') as openverse_tag), array[1]))
+            end) inaturalist_name_tag,
+        tags_nonenglish_vernacular
+    from inaturalist.taxa
+    LEFT JOIN catalog_of_life
+        on (cast(md5(taxa.name) as uuid) = catalog_of_life.md5_scientificname)
+    LEFT JOIN inaturalist.manual_name_additions
+        on (cast(md5(taxa.name) as uuid) = manual_name_additions.md5_scientificname)
+    where taxa.name <> 'Not assigned'
+);
+ALTER TABLE inaturalist.taxa_with_vernacular ADD PRIMARY KEY (taxon_id);
+COMMIT;
+
+/*
+           ********** Create enriched table with ancestry tags *********
+
+Join each record to all of its ancestor records and aggregate ancestor titles into the
+tags along with json strings from the enriched data above.
+    + expand each ancestry string into an array
+    + get the taxa record for each value of the array
+    + aggregate back to the original taxon level, with tags for ancestor names (titles)
```

Comment:
Would you be able to provide some examples of data at each of these stages? E.g. `expand each ancestry string into an array ('reptilia/crocodilia/somethingelseia' -> ['reptilia', 'crocodilia', 'somethingelsia'])`.

Reply:
Yes. Does the example below ([here](https://github.com/WordPress/openverse-catalog/blob/4f01d3b0378f2f0e33e51ebf92c1a3bcea32cd28/openverse_catalog/dags/providers/provider_csv_load_scripts/inaturalist/taxa.sql#L147)) help?